src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,276 @@
+import org.whispersystems.textsecuregcm.util.Util;
+import redis.clients.jedis.Jedis;
+
+import javax.ws.rs.ProcessingException;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import java.util.Random;
+
+import static com.codahale.metrics.MetricRegistry.name;
+
+public class DirectoryReconciler implements Managed {
+
+  private static final Logger logger = LoggerFactory.getLogger(DirectoryReconciler.class);
+
+  private static final MetricRegistry metricRegistry       = SharedMetricRegistries.getOrCreate(Constants.METRICS_NAME);
+  private static final Timer          readBucketTimer      = metricRegistry.timer(name(DirectoryReconciler.class, "readBucket"));
+  private static final Timer          sendBucketTimer      = metricRegistry.timer(name(DirectoryReconciler.class, "sendBucket"));
+  private static final Meter          sendBucketErrorMeter = metricRegistry.meter(name(DirectoryReconciler.class, "sendBucketError"));
+
+  private static final String   WORKERS_KEY   = "directory_reconciliation_workers";
+  private static final String   COUNTER_KEY   = "directory_reconciliation_counter";
+  private static final long     WORKER_TTL_MS = 120_000L;
+  private static final long     BUCKET_ORDER  = 11L;
+  private static final long     BUCKET_COUNT  = 1L << BUCKET_ORDER;
+  private static final double   JITTER_MAX    = 0.20;
+  private static final String[] COUNTER_ARGS  = {
+          "OVERFLOW", "WRAP",
+          "INCRBY", "u32", "0", "1",
+          };
+
+  private final String              serverApiUrl;
+  private final ReplicatedJedisPool jedisPool;
+  private final AccountsManager     accountsManager;
+  private final Client              client;
+
+  private DirectoryReconciliationWorker directoryReconciliationWorker;
+
+  public DirectoryReconciler(ContactDiscoveryConfiguration cdsConfig,
+                             ReplicatedJedisPool jedisPool,
+                             AccountsManager accountsManager)
+  {
+    this.serverApiUrl = cdsConfig.getServerApiUrl();
+    this.jedisPool = jedisPool;
+    this.accountsManager = accountsManager;
+
+    SslConfigurator sslConfig = SslConfigurator.newInstance()
+                                               .securityProtocol("TLSv1.2");
+    try {
+      sslConfig.trustStore(initializeKeyStore(cdsConfig.getServerApiCaCertificate()));
+    } catch (CertificateException ex) {
+      logger.error("error reading serverApiCaCertificate from contactDiscovery config", ex);
```

Comment:
shouldn't the constructor throw this? I would assume we want to interrupt startup if we can't configure the client correctly